cpu/nrf5x: implement wdt peripheral driver

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR implement the watchdog peripheral interface as proposed in #11527 on nRF5x CPU.

It has been tested with success on nrf52832-mdk, nrf52840-mdk and microbit.

On nrf, it's not possible to stop the watchdog once it's started so when setting up a timeout callback, the CPU is reset in all the case after 2 cycles of 32768kHz clock. This is quite fast and doesn't allow complex task to happen before reset: so I'm wondering if it's relevant to provide `periph_wdt_cb` on this platform.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

I used the following sample application to test the feature but it'll be more useful to test with the test application provided in #11252:
- main.c
```c
#include <stdio.h>

#include "xtimer.h"
#include "periph/wdt.h"

int main(void)
{
    puts("Setup watchdog");
    wdt_setup_reboot(0, 2000);
    wdt_start();

    uint8_t cnt = 5;
    printf("%d\n", cnt);
    while (cnt--) {
        xtimer_sleep(1);
        printf("%d\n", cnt);
        wdt_kick();
    }

    puts("Reboot");

    return 0;
}
```
- Makefile:
```mk
APPLICATION = wdt-sample
BOARD ?= nrf52830-mdk
RIOTBASE ?= $(CURDIR)/../..

FEATURES_REQUIRED += periph_wdt
USEMODULE += xtimer

include $(RIOTBASE)/Makefile.include
```

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

Based on #11527 

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

benpicco commented on cpu/nrf5x_common/periph/wdt.c:
```diff
@@ -0,0 +1,116 @@
+void wdt_setup_callback(uint32_t min_time, uint32_t max_time,
+                        wdt_cb_t cb, void* arg)
+{
+    wdt_cb = cb;
```
What if `cb == NULL`?

aabadie commented:
Rebased and comment addressed @benpicco. Not sure how to handle WDT_WARNING_PERIOD on this platform since it's less than a ms (e.g. 2 32.768kHz clock cycles) and not configurable.

benpicco commented:
This will still crash if you call `wdt_setup_reboot_with_callback(0, 123, NULL, NULL);`.

AFAIU `WDT_WARNING_PERIOD` is only a suggestion. If you can't implement it, just ignore it.

aabadie commented:
> This will still crash if you call wdt_setup_reboot_with_callback(0, 123, NULL, NULL);

I just forgot to push...

> AFAIU WDT_WARNING_PERIOD is only a suggestion. If you can't implement it, just ignore it.

Ok. It can be implemented but no complex actions can be performed before reset. I think some GPIO could be set/cleared before reset but no logging on UART is possible for example.

benpicco commented on cpu/nrf5x_common/periph/wdt.c:
```diff
@@ -0,0 +1,119 @@
+        /* enable interrupt */
+        NVIC_EnableIRQ(WDT_IRQn);
+        NRF_WDT->INTENSET = WDT_INTENSET_TIMEOUT_Enabled;
+    }
```
```suggestion
    } else {
        NVIC_DisableIRQ(WDT_IRQn);
    }
```

aabadie replied:
I can't find a use case for this. The setup function should be called only once at startup and the watchdog can be started. On nrf, it's not possible to stop a running watchdog and it's also not possible to reconfiguring after it has been started.
The only case I can see is when an application is configuring it several, differently, before starting it. And this is pretty useless I think.

benpicco replied:
This wouldn't disable the watchdog though, it would just disable the callback.

aabadie replied:
Ok I changed it without the else but rather disable before eventually enable it if the callback is not null.

benpicco commented:
I tested with my [`tests/periph_wdt`](https://github.com/RIOT-OS/RIOT/blob/4d98bf39d143c13e7c717680d64689ca832ee73c/tests/periph_wdt/main.c) and basic watchdog operation is working well.

I had to disable the window mode test as the assert would fail. (Maybe we should introduce `FEATURES_PROVIDED += periph_wdt_window` too)

> On nrf, it's not possible to stop the watchdog once it's started so when setting up a timeout callback, the CPU is reset in all the case after 2 cycles of 32768kHz clock.

I thought this would still be enough to call `wdt_kick()`, but this doesn't seem to be the case. Even when putting `NRF_WDT->RR[0] = WDT_RR_RR_Reload;` directly into `isr_wdt()` tha board would reset.

So I wonder what the callback can be used for.

aabadie commented:
> I wonder what the callback can be used for.

Toggle a pin ?

benpicco commented:
I tried enabling a LED but I guess it would reset too fast to see it (it's only 61µs after all) and I didn't get a scope to verify it.

But this can still be useful. Please add a comment to `wdt_setup_reboot_with_callback` that
 - the MCU will reset within 2 cycles of the 32768 Hz clock (61µs)
 - the reset can not be stopped

Then this should be good to go.

aabadie commented:
Thanks for having a look at this PR @benpicco.

> Then this should be good to go.

I would like to have a test for this feature merged before. I think the one in #11252 would be good and I'd like to merge this first.

aabadie commented:
Rebased, now that #11252 is merged.

I also adapted slightly when printing the range values in the test application: since upper limit is an uint32_t it can potentially have very large values (even it's pointless). I added a couple commits about that.
In the nrf implementation, there's nothing that prevent very large values so the upper limit is set as the maximum possible in the uint32_t range.

fjmolinas commented on cpu/stm32_common/include/periph_cpu_common.h:
```diff
@@ -104,7 +104,7 @@ extern "C" {
 /* Actual Lower Limit is ~100us so round up */
 #define NWDT_TIME_LOWER_LIMIT          (1U)
-#define NWDT_TIME_UPPER_LIMIT          (4U * US_PER_MS * 4096U * (1 << 6U) \
+#define NWDT_TIME_UPPER_LIMIT          (uint32_t)(4U * US_PER_MS * 4096U * (1 << 6U) \
```
Isn't `U` for unsigned int? which if I'm not mistake us 32bit for arm-gcc, and the other variables get casted to `U` as well.

aabadie replied:
I reverted that change and now only cast in the test application.

fjmolinas commented on cpu/nrf5x_common/include/periph_cpu_common.h:
```diff
@@ -165,6 +165,14 @@ typedef struct {
+ * @{
+ */
+#define NWDT_TIME_LOWER_LIMIT          (1)
+#define NWDT_TIME_UPPER_LIMIT          (UINT32_MAX + 1 / 32.768)
```
Can you explain this please.

aabadie replied:
There was a mistake, I pushed a fix and added a comment.

benpicco replied:
Is this really a floating point number that you want to use as divisor?

aabadie replied:
Good point. I'll change that

benpicco commented on cpu/nrf5x_common/periph/wdt.c:
```diff
@@ -87,7 +87,7 @@ void wdt_setup_reboot(uint32_t min_time, uint32_t max_time)
 
     /* timeout (s) = (CRV + 1) / 32768 */
-    uint32_t crv = (uint32_t)(max_time * 32.768) - 1;
+    uint32_t crv = (uint32_t)((max_time << 15) / 1000) - 1;
```
Now you don't need that cast anymore.

aabadie replied:
Doesn't work if I change that.

aabadie replied:
ping @benpicco

benpicco replied:
I'm not in office right now and don't have any Nordic devices with me - will test again next week.

fjmolinas commented:
Tested on `nrf52840-mdk`:

<details><summary>make -C tests/periph_wdt/ BOARD=nrf52840-mdk flash test</summary>

```

make: Entering directory '/home/francisco/workspace/RIOT/tests/periph_wdt'
Building application "tests_periph_wdt" for "nrf52840-mdk" with MCU "nrf52".

"make" -C /home/francisco/workspace/RIOT/boards/nrf52840-mdk
"make" -C /home/francisco/workspace/RIOT/core
"make" -C /home/francisco/workspace/RIOT/cpu/nrf52
"make" -C /home/francisco/workspace/RIOT/cpu/cortexm_common
"make" -C /home/francisco/workspace/RIOT/cpu/cortexm_common/periph
"make" -C /home/francisco/workspace/RIOT/cpu/nrf52/periph
"make" -C /home/francisco/workspace/RIOT/cpu/nrf5x_common
"make" -C /home/francisco/workspace/RIOT/cpu/nrf5x_common/periph
"make" -C /home/francisco/workspace/RIOT/drivers
"make" -C /home/francisco/workspace/RIOT/drivers/periph_common
"make" -C /home/francisco/workspace/RIOT/sys
"make" -C /home/francisco/workspace/RIOT/sys/auto_init
"make" -C /home/francisco/workspace/RIOT/sys/div
"make" -C /home/francisco/workspace/RIOT/sys/isrpipe
"make" -C /home/francisco/workspace/RIOT/sys/newlib_syscalls_default
"make" -C /home/francisco/workspace/RIOT/sys/shell
"make" -C /home/francisco/workspace/RIOT/sys/stdio_uart
"make" -C /home/francisco/workspace/RIOT/sys/tsrb
"make" -C /home/francisco/workspace/RIOT/sys/xtimer
   text    data     bss     dec     hex filename
  12132     136    2664   14932    3a54 /home/francisco/workspace/RIOT/tests/periph_wdt/bin/nrf52840-mdk/tests_periph_wdt.elf
/home/francisco/workspace/RIOT/dist/tools/pyocd/pyocd.sh flash /home/francisco/workspace/RIOT/tests/periph_wdt/bin/nrf52840-mdk/tests_periph_wdt.hex
### Flashing Target ###
[====================] 100%
0001880:INFO:loader:Erased 12288 bytes (3 sectors), programmed 12288 bytes (3 pages), skipped 0 bytes (0 pages) at 9.77 kB/s
Done flashing
PASSED target time: 100.0[ms], actual_time: 100.464[ms]
PASSED target time: 500.0[ms], actual_time: 498.508[ms]
PASSED target time: 1000.0[ms], actual_time: 999.033[ms]
PASSED target time: 5000.0[ms], actual_time: 4991.131[ms]
TEST PASSED

make: Leaving directory '/home/francisco/workspace/RIOT/tests/periph_wdt'
```
</details>

<details><summary>make -C tests/periph_wdt/ BOARD=nrf52840dk flash test</summary>

```
make: Entering directory '/home/francisco/workspace/RIOT/tests/periph_wdt'
Building application "tests_periph_wdt" for "nrf52840dk" with MCU "nrf52".

"make" -C /home/francisco/workspace/RIOT/boards/nrf52840dk
"make" -C /home/francisco/workspace/RIOT/boards/common/nrf52xxxdk
"make" -C /home/francisco/workspace/RIOT/core
"make" -C /home/francisco/workspace/RIOT/cpu/nrf52
"make" -C /home/francisco/workspace/RIOT/cpu/cortexm_common
"make" -C /home/francisco/workspace/RIOT/cpu/cortexm_common/periph
"make" -C /home/francisco/workspace/RIOT/cpu/nrf52/periph
"make" -C /home/francisco/workspace/RIOT/cpu/nrf5x_common
"make" -C /home/francisco/workspace/RIOT/cpu/nrf5x_common/periph
"make" -C /home/francisco/workspace/RIOT/drivers
"make" -C /home/francisco/workspace/RIOT/drivers/periph_common
"make" -C /home/francisco/workspace/RIOT/sys
"make" -C /home/francisco/workspace/RIOT/sys/auto_init
"make" -C /home/francisco/workspace/RIOT/sys/div
"make" -C /home/francisco/workspace/RIOT/sys/isrpipe
"make" -C /home/francisco/workspace/RIOT/sys/newlib_syscalls_default
"make" -C /home/francisco/workspace/RIOT/sys/shell
"make" -C /home/francisco/workspace/RIOT/sys/stdio_uart
"make" -C /home/francisco/workspace/RIOT/sys/tsrb
"make" -C /home/francisco/workspace/RIOT/sys/xtimer
   text    data     bss     dec     hex filename
  12252     136    2672   15060    3ad4 /home/francisco/workspace/RIOT/tests/periph_wdt/bin/nrf52840dk/tests_periph_wdt.elf
/home/francisco/workspace/RIOT/dist/tools/jlink/jlink.sh flash /home/francisco/workspace/RIOT/tests/periph_wdt/bin/nrf52840dk/tests_periph_wdt.bin
### Flashing Target ###
### Flashing at base address 0x0 with offset 0 ###
SEGGER J-Link Commander V6.46g (Compiled Jun 14 2019 19:36:40)
DLL version V6.46g, compiled Jun 14 2019 19:36:30

J-Link Commander will now exit on Error

J-Link Command File read successfully.
Processing script file...

J-Link connection not established yet but required for command.
Connecting to J-Link via USB...O.K.
Firmware: J-Link OB-SAM3U128-V2-NordicSemi compiled Jan  7 2019 14:07:15
Hardware version: V1.00
S/N: 683890121
VTref=3.300V
Target connection not established yet but required for command.
Device "NRF52" selected.


Connecting to target via SWD
Found SW-DP with ID 0x2BA01477
Found SW-DP with ID 0x2BA01477
Scanning AP map to find all available APs
AP[2]: Stopped AP scan as end of AP map has been reached
AP[0]: AHB-AP (IDR: 0x24770011)
AP[1]: JTAG-AP (IDR: 0x02880000)
Iterating through AP map to find AHB-AP to use
AP[0]: Core found
AP[0]: AHB-AP ROM base: 0xE00FF000
CPUID register: 0x410FC241. Implementer code: 0x41 (ARM)
Found Cortex-M4 r0p1, Little endian.
FPUnit: 6 code (BP) slots and 2 literal slots
CoreSight components:
ROMTbl[0] @ E00FF000
ROMTbl[0][0]: E000E000, CID: B105E00D, PID: 000BB00C SCS-M7
ROMTbl[0][1]: E0001000, CID: B105E00D, PID: 003BB002 DWT
ROMTbl[0][2]: E0002000, CID: B105E00D, PID: 002BB003 FPB
ROMTbl[0][3]: E0000000, CID: B105E00D, PID: 003BB001 ITM
ROMTbl[0][4]: E0040000, CID: B105900D, PID: 000BB9A1 TPIU
ROMTbl[0][5]: E0041000, CID: B105900D, PID: 000BB925 ETM
Cortex-M4 identified.
Halting CPU for downloading file.
Downloading file [/home/francisco/workspace/RIOT/tests/periph_wdt/bin/nrf52840dk/tests_periph_wdt.bin]...
Comparing flash   [100%] Done.
Erasing flash     [100%] Done.
Programming flash [100%] Done.
Verifying flash   [100%] Done.
J-Link: Flash download: Bank 0 @ 0x00000000: 1 range affected (16384 bytes)
J-Link: Flash download: Total time needed: 0.622s (Prepare: 0.042s, Compare: 0.004s, Erase: 0.337s, Program: 0.228s, Verify: 0.000s, Restore: 0.008s)
O.K.

Reset delay: 0 ms
Reset type NORMAL: Resets core & peripherals via SYSRESETREQ & VECTRESET bit.
Reset: Halt core after reset via DEMCR.VC_CORERESET.
Reset: Reset device via AIRCR.SYSRESETREQ.



Script processing completed.

PASSED target time: 100.0[ms], actual_time: 100.463[ms]
PASSED target time: 500.0[ms], actual_time: 502.544[ms]
PASSED target time: 1000.0[ms], actual_time: 1005.088[ms]
PASSED target time: 5000.0[ms], actual_time: 5025.442[ms]
TEST PASSED

make: Leaving directory '/home/francisco/workspace/RIOT/tests/periph_wdt'
```

</details>

fjmolinas commented:
@benpicco are you ok with the current status? Shouls @aabadie squash?

benpicco commented:
Yes, please go ahead and squash - the driver looks good.
I just read something didn't work when removing the float use, but it looks like @aabadie figured it out :)

aabadie commented:
squash!

aabadie commented:
@benpicco, are you ok to merge this one too ?

aabadie commented:
To be sure there's no problem, I rebased on top of master now that #11520 in merged.
